### ISSUE TITLE
force_ssl breaks Amazon ELB healthchecks

### ISSUE THREAD

**Issue by jf647** [chef/supermarket#986]
The attributes file for the supermarket cookbook says:

```
# To use AWS ELB in front of Supermarket, set force_ssl to true, but
# don't set the ssl_crt_path or ssl_key_path. To use SSL directly on
# the nginx proxy in front of the app, set force_ssl to true and set
# the path attributes below.
default['supermarket']['force_ssl']  = false
```

This might work if you just have an ELB, but it completely breaks if you have a healthcheck attached to that ELB hitting the http port.

When you have SSL termination on the ELB hitting HTTP on the supermarket box, the ELB will deliver proxied request with the X-Forwarded-Proto header set to 'https'.

Healthchecks (because they originate on the ELB and are not made on anyone's behalf) do not have this header.  So an attempt to hit the healthcheck URL /api/v1/health is returned as 301 redirect to the https version of the site.

ELB healthchecks consider something healthy only if they get a 2xx (or maybe even only 200 - not entierly sure).  They definitely don't follow or like 3xx series responses.  Following the redirect wouldn't even make sense because that would result in the ELB hitting itself rather than the supermarket instance.

Net effect is that following the instructions in the attribute file result in an ELB that never considers the supermarket instance to be in service.

**Comment by jf647** [comment]
We are considering a PR to fix this along these lines:

Create a few new attributes:

* don't redirect if header set (boolean)
* header value to match to disable redirection
* load balancer type

The load balancer type would be 'elb' ,'bigip', 'haproxy', 'varnish', etc.  If set, it would set the header value to match.  I can only speak to ELBs, but they set this header:

    User-Agent: ELB-HealthChecker/1.0

Others can contribute the correct strings for other load balancer types.  You can always skip the load balancer type and just set the header to match directly if you want.

Inside the nginx template, if the don't redirect attribute is true and the header matches, then the redirection won't happen.

Would appreciate some comment from the maintainers of the cookbook before moving forward.  @cwebberOps ?

**Comment by smith** [comment]
This looks decent, but I'm wary about having the configuration have too much knowledge about the load balancer it's using. Conceptually, it shouldn't matter.

What if there was some kind of `health_check_user_agent` attribute that, if set, would behave as you are describing? We can document common ones and would have a more generic, feature testing solution rather than having to change something every time somebody tries to run with a new kind of LB.

**Comment by jf647** [comment]
That sounds fine, it's basically what I was referring to when I said "You can always skip the load balancer type and just set the header to match directly if you want."  The predefined map from common types to user agents (in code rather than in docs) was only for convenience.

**Comment by nellshamrell** [comment]
This was fixed with Supermarket 2.0.0, which allows a user to set [‘ssl’][‘enabled’] to false.
